File: 2641-cousins-in-binary-tree-ii/2641-cousins-in-binary-tree-ii.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

#include <stdio.h>
#include <stdlib.h>

/**
 * @brief Replace each node value with the sum of its cousins' values in the binary tree.
 *
 * This function modifies the binary tree by replacing the value of each node with
 * the sum of all its cousins' values. Two nodes are cousins if they are at the same
 * depth but have different parents.
 *
 * @param root (struct TreeNode*): The root of the binary tree.
 * 
 * @return struct TreeNode*: The modified binary tree with updated node values.
 */
struct TreeNode* replaceValueInTree(struct TreeNode* root) {
    if (!root) {
        return NULL;
    }

    // Define the queue for level order traversal (BFS)
    struct TreeNode** queue = (struct TreeNode**)malloc(100000 * sizeof(struct TreeNode*));
    int front = 0, rear = 0;
    
    // Initialize queue with the root node and set root value to 0 since it has no cousins
    queue[rear++] = root;
    root->val = 0;  // The root node has no cousins, so set its value to 0

    // Loop through the tree using breadth-first search (BFS)
    while (front < rear) {
        int level_size = rear - front;  // Number of nodes at the current level
        int level_sum = 0;  // Sum of all nodes' values at the current level
        
        // Temporary array to hold nodes for this level
        struct TreeNode** temp_nodes = (struct TreeNode**)malloc(level_size * sizeof(struct TreeNode*));
        
        // First pass: calculate the sum of the current level and store nodes in the temp array
        for (int i = 0; i < level_size; i++) {
            struct TreeNode* node = queue[front++];
            temp_nodes[i] = node;
            
            // Add the children to the queue for the next level
            if (node->left) {
                queue[rear++] = node->left;
                level_sum += node->left->val;
            }
            if (node->right) {
                queue[rear++] = node->right;
                level_sum += node->right->val;
            }
        }
        
        // Second pass: update node values based on the cousin sum
        for (int i = 0; i < level_size; i++) {
            struct TreeNode* node = temp_nodes[i];
            int sibling_sum = 0;

            // Calculate the sibling sum (sum of children of the current node)
            if (node->left) {
                sibling_sum += node->left->val;
            }
            if (node->right) {
                sibling_sum += node->right->val;
            }

            // Update the values of the left and right children with the cousin sum
            if (node->left) {
                node->left->val = level_sum - sibling_sum;
            }
            if (node->right) {
                node->right->val = level_sum - sibling_sum;
            }
        }
        
        // Free temporary storage
        free(temp_nodes);
    }

    // Free the queue used for BFS
    free(queue);
    
    return root;
}

File: 2641-cousins-in-binary-tree-ii/2641-cousins-in-binary-tree-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def sumup(self, root, sums, lvl):
        """
        Recursively sums up node values at each level.

        Parameters:
        root (TreeNode): The current node.
        sums (list): List to store the sum of node values at each level.
        lvl (int): The current level in the tree.
        """
        if not root:
            return

        # Ensure sums list is long enough to accommodate the current level
        if lvl >= len(sums):
            sums.append(0)

        sums[lvl] += root.val
        self.sumup(root.left, sums, lvl + 1)
        self.sumup(root.right, sums, lvl + 1)

    def replace(self, root, sums, lvl):
        """
        Recursively replaces node values with their cousin sums.

        Parameters:
        root (TreeNode): The current node.
        sums (list): List storing the sum of node values at each level.
        lvl (int): The current level in the tree.
        """
        if not root:
            return

        # Recursively process left and right children
        self.replace(root.left, sums, lvl + 1)
        self.replace(root.right, sums, lvl + 1)

        # Calculate sibling sum (sum of the current node's children if they exist)
        left_val = root.left.val if root.left else 0
        right_val = root.right.val if root.right else 0

        # Replace the left child's value if it exists
        if root.left:
            root.left.val = sums[lvl + 1] - left_val - right_val

        # Replace the right child's value if it exists
        if root.right:
            root.right.val = sums[lvl + 1] - right_val - left_val

    def replaceValueInTree(self, root):
        """
        Main function to replace node values in the tree with the sum of cousin node values.

        Parameters:
        root (TreeNode): The root of the binary tree.

        Returns:
        TreeNode: The modified tree with updated node values.
        """
        sums = []  # List to store sums of each level

        # First pass: Calculate the sum of values at each level
        self.sumup(root, sums, 0)

        # Second pass: Replace node values with the cousin sums
        self.replace(root, sums, 0)

        # The root node has no cousins, so set its value to 0
        root.val = 0

        return root